src/charm.py

```diff
@@ -414,6 +418,48 @@ def _on_certificate_creation_request(self, event: CertificateCreationRequestEven
                 relation_id=event.relation_id,
             )
 
+    def _on_generate_certificate_action(self, event: ActionEvent) -> None:
+        """Generates TLS Certificate.
+
+        Generates a private key and certificate for an external service.
+        Args:
+            event: Juju event.
+
+        Returns:
+            None
+        """
+        if not self.unit.is_leader():
+            event.fail(message="Action cannot be run on non-leader unit")
+            return
+
+        sans = None
```

Comment:
This block should be computed after the `if self._self_signed_certificates` since there's no need to compute it if it's not self-signed.

Reply:
Done, moved the check to the top